src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -127,5 +128,19 @@ internal static MultiClusterRegistrationStrategy GetMultiClusterRegistrationStra
                             grainClass.Name));
             }
         }
+
+        // used by Orleankka's bootstrapper
+        internal void SetMessageAlwaysInterleaves(Func<object, bool> predicate)
+        {
+            if (predicate == null)
+                throw new ArgumentNullException(nameof(predicate));
+
+            IsMessageInterleavesPredicate = predicate;
+        }
+
+        public bool IsMessageInterleaves(Message message)
+        {
+            return IsReentrant || IsMessageInterleavesPredicate(message.BodyObject.GetType());
```

Comment:
Please add 2 tests:
one with successful interleaving and one with callback throwing exception . I want to see that the exception is handled correctly. That will only make us define what "correctly" is in this case.